### PR TITLE
Enable system metrics logging for resuming an existing run

### LINKED ISSUE #10253: [BUG] "log_system_metrics" not working when "existing_run_id" is set
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

- Client: 2.8.0
- Tracking server: 2.8.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 22.04
- **Python version**: 3.10.12

### Issue
When calling `mlflow.start_run` with `log_system_metrics=True` and `run_id` set by parameter or environment variable, the `SystemMetricsMonitor` is not going to be started. System metrics are only recorded for newly created runs, while they are not for continuing existing runs. This is a bug, since one would expect that system metrics are recorded in both cases.

This bug might be unintended due to an incorrect indentation of the code block in `mlflow/tracking/fluent.py` which starts the monitoring process only in the else-branch (i.e. when `existing_run_id` is not set).
image

It might also be intended in order to avoid dealing with the question from which `step` the recording of the metrics should continue for existing runs. When moving the code block outside of the else-branch, system metrics are going to be logged for existing runs with the current timestamp, but starting from step 0 again. Hence, the plot with the time on the x-axis is still correct, but not the plot with the steps on the x-axis.
image
image

There is also a minor bug in `system_metrics_monitor.py`, which I have found when inspecting the other issue. Since the system metrics are logged with `BatchMetricsLogger`, it is advisable to call `self.mlflow_logger.flush` when stopping the monitoring process.

### Proposed Solution
My suggestion is to move the code block (see above) out of the else-branch in any case, as it still provides reasonable plots with the time on the x-axis and having the issue with the steps on the x-axis is acceptable for an experimental feature in my opinion.

For solving the issue with the steps on the x-axis, I can think of different solutions for setting `_logging_step` in the constructor of `SystemMetricsLogger`:
- determine the largest step for one of the system metrics from which the logging should be continued using  `mlflow.client.MlflowClient().get_metric_history(<run_id>, <metric_key>)` (The question is which metric should be taken? In particular, it has to be one which is always logged, i.e. not `system/gpu...`.)
- determine the largest step among all system metrics from which the logging should be continued (This is more robust, in case some system metrics are not logged and some are. However, it might be inefficient to iterate through all metric histories...)
- log an additional metric `system/_logging_step` from which the logging should be continued (Not my favorite one...)

One can also think of extending this to have better looking plots:
- add an offset of X steps when continuing a run in order to artificially create some distance between the two runs in the plots
- log a NaN value first when continuing a runs in order to artificially create a gap between the two runs in the plot

But here I need more input from the maintainers/community regarding the design of a fix.

Regarding the minor issue, I would simply add `self.mlflow_logger.flush()` to the end of `SystemMetricsMonitor.finish` once all monitoring processes have been stopped.

PS: That's my first Github activity, so I hope I did everything correct:D Let me know if I should provide more details/do something differently.

### Tracking information

```shell
System information: Linux #1 SMP Fri Jan 27 02:56:13 UTC 2023
Python version: 3.10.12
MLflow version: 2.8.0
MLflow module location: x
Tracking URI: x
Registry URI: x
MLflow dependencies: 
  Flask: 2.3.3
  Jinja2: 3.1.2
  alembic: 1.12.0
  click: 8.1.7
  cloudpickle: 2.2.1
  databricks-cli: 0.18.0
  docker: 6.1.3
  entrypoints: 0.4
  gitpython: 3.1.40
  gunicorn: 21.2.0
  importlib-metadata: 6.8.0
  markdown: 3.5
  matplotlib: 3.8.0
  numpy: 1.25.2
  packaging: 23.2
  pandas: 2.1.1
  protobuf: 4.24.4
  psutil: 5.9.6
  pyarrow: 13.0.0
  pytz: 2023.3.post1
  pyyaml: 6.0.1
  querystring-parser: 1.2.4
  requests: 2.31.0
  scikit-learn: 1.3.1
  scipy: 1.11.3
  sqlalchemy: 2.0.22
  sqlparse: 0.4.4
```

### Code to reproduce issue

```
import mlflow
import time

mlflow.set_system_metrics_sampling_interval(1)

with mlflow.start_run(log_system_metrics=True) as run:
    time.sleep(2)

with mlflow.start_run(log_system_metrics=True, run_id=run.info.run_id):
    time.sleep(2)
```

### Stack trace

*This bug does not generate an exception, but it will become visible in the log (see below).*

### Other info / logs

Expected output:
```
2023/10/31 18:53:45 INFO mlflow.system_metrics.system_metrics_monitor: Started monitoring system metrics.
2023/10/31 18:53:47 INFO mlflow.system_metrics.system_metrics_monitor: Stopping system metrics monitoring...
2023/10/31 18:53:47 INFO mlflow.system_metrics.system_metrics_monitor: Successfully terminated system metrics monitoring!
2023/10/31 18:53:47 INFO mlflow.system_metrics.system_metrics_monitor: Started monitoring system metrics.
2023/10/31 18:53:49 INFO mlflow.system_metrics.system_metrics_monitor: Stopping system metrics monitoring...
2023/10/31 18:53:49 INFO mlflow.system_metrics.system_metrics_monitor: Successfully terminated system metrics monitoring!
```

Actual output:
```
2023/10/31 18:53:45 INFO mlflow.system_metrics.system_metrics_monitor: Started monitoring system metrics.
2023/10/31 18:53:47 INFO mlflow.system_metrics.system_metrics_monitor: Stopping system metrics monitoring...
2023/10/31 18:53:47 INFO mlflow.system_metrics.system_metrics_monitor: Successfully terminated system metrics monitoring!
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by chenmoneygithub**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/chenmoneygithub/mlflow/pull/10312?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10312/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10312
```

</p>
</details>

### Related Issues/PRs

 

Resolve #10253 

### What changes are proposed in this pull request?

Enable system metrics logging for resuming an existing run.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/system_metrics/system_metrics_monitor.py [hunk 1]
```diff
@@ -134,6 +143,7 @@ def finish(self):
         """Stop monitoring system metrics."""
         if self._process is None:
             return
+        self.mlflow_logger.flush()
```
Comment (danielyxyang) [inline mlflow/system_metrics/system_metrics_monitor.py:146]: Wouldn't it be better to flush at the very end of the `finish` function, once all monitoring processes have been stopped? Because in the very unlikely case, some process might record new metrics between flushing the metrics and stopping the processes😅

File: mlflow/system_metrics/system_metrics_monitor.py [hunk 2]
```diff
@@ -60,7 +69,7 @@ def __init__(self, run_id, sampling_interval=10, samples_before_logging=1):
         self.mlflow_logger = BatchMetricsLogger(self._run_id)
         self._shutdown_event = threading.Event()
         self._process = None
-        self._logging_step = 0
+        self._logging_step = initial_logging_step
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply mlflow/system_metrics/system_metrics_monitor.py:72]: > The user might have a hard time to figure out how to find the last logging step

If users want to customize their system metrics logging by instantiating `SystemMetricsMonitor`, and they want to use a different `initial_logging_step`, it shouldn't be a block for them. 

> suggest to set initial_logging_step to None or "auto" by default

I am not a fan of `None`/`auto` as it's unclear what really happens inside the constructor.

> I prefer having the logic to be as close to where this key is defined. 

It's always a problem. Putting it closer won't solve the issue. But yes i don't like the hard-coded string here, I will make a util function for it.

File: mlflow/system_metrics/system_metrics_monitor.py [hunk 3]
```diff
@@ -60,8 +69,26 @@ def __init__(self, run_id, sampling_interval=10, samples_before_logging=1):
...
+        client = MlflowClient()
+        try:
+            run = client.get_run(run_id)
+        except Exception:
+            return 0
+        system_metric_name = None
+        for metric_name in run.data.metrics.keys():
+            if metric_name.startswith(self._metrics_prefix):
+                system_metric_name = metric_name
+                break
```
Comment (harupy) [inline mlflow/system_metrics/system_metrics_monitor.py:87]: Not sure if you're a fan of this syntax, but you can use for-else:

```suggestion
        for metric_name in run.data.metrics.keys():
            if metric_name.startswith(self._metrics_prefix):
                system_metric_name = metric_name
                break
        else:
             return 0
```

File: mlflow/tracking/fluent.py [hunk 4]
```diff
@@ -376,20 +376,25 @@ def start_run(
...
-                run_id_to_system_metrics_monitor[active_run_obj.info.run_id] = system_monitor
-                system_monitor.start()
-            except Exception as e:
-                _logger.error(f"Failed to start system metrics monitoring: {e}.")
+
+    if log_system_metrics is None:
+        # if `log_system_metrics` is not specified, we will check environment variable.
+        log_system_metrics = MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING.get()
+    if log_system_metrics:
+        resume_logging = bool(existing_run_id)
```
Comment (harupy) [inline mlflow/tracking/fluent.py:384]:
```suggestion
        resume_logging = existing_run_id is None
```

File: mlflow/system_metrics/system_metrics_monitor.py (hunk 3 above)
Comment (chenmoneygithub) [reply mlflow/system_metrics/system_metrics_monitor.py:87]: ohhh... not a fan - -"

File: mlflow/tracking/fluent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielyxyang) [reply mlflow/tracking/fluent.py:384]: I would also directly write it inline, saves one line and maybe a bit more readable I guess...:wink: 
```python
system_monitor = SystemMetricsMonitor(
    active_run_obj.info.run_id,
    resume_logging=existing_run_id is not None,
)
```

**Review by danielyxyang — CHANGES_REQUESTED** [review]

File: mlflow/tracking/fluent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/tracking/fluent.py:384]: Sorry, I meant `existing_run_id is not None`.

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/tracking/fluent.py:384]: +1 to https://github.com/mlflow/mlflow/pull/10312/files#r1390278676

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/tracking/fluent.py:384]: You could still argue that `existing_run_id` is always None or a truthy value, but I'd still prefer `is None`.

File: mlflow/system_metrics/system_metrics_monitor.py [hunk 5]
```diff
@@ -60,8 +69,26 @@ def __init__(self, run_id, sampling_interval=10, samples_before_logging=1):
...
+        if system_metric_name is None:
+            return 0
+        metric_history = client.get_metric_history(run_id, system_metric_name)
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply mlflow/system_metrics/system_metrics_monitor.py:90]: > which basically only assumes that CPU metrics are always logged
That's a fragile assumption as well, and I don't like hard-coded strings.

> This code also assumes that all system metrics are logged in lockstep
You will have the same problem for CPU metrics logging.

**Comment by chenmoneygithub** [comment]
@danielyxyang Thanks for your help! But please don't start contributing by doing code review. Usually we assign reviews code review to OSS contributors after they have nailed several solid PRs, otherwise the review could be distracting and increasing our overhead.

**Comment by danielyxyang** [comment]
Oops sorry for that! I'll let you do your work then:)

File: docs/source/getting-started/quickstart-2/index.rst [hunk 6]
```diff
@@ -280,5 +323,7 @@ In addition, some cloud providers have built-in support for MLflow. For instance
 - `Amazon SageMaker <https://docs.aws.amazon.com/sagemaker/index.html>`_
 - `Google Cloud <https://cloud.google.com/doc>`_
 
-all support MLflow. Cloud platforms generally support multiple workflows for deployment: command-line, SDK-based, and Web-based. You can use MLflow in any of these workflows, although the details will vary between platforms and versions. Again, you will need to consult your cloud provider's documentation for details.
+all support MLflow. Cloud platforms generally support multiple workflows for deployment: command-line,
```
Comment (harupy) [inline docs/source/getting-started/quickstart-2/index.rst:326]: Is this change related to this PR?

File: tests/system_metrics/test_system_metrics_logging.py [hunk 7]
```diff
@@ -91,3 +91,52 @@ def test_automatic_system_metrics_monitor():
...
+        time.sleep(2)
+
+    # Pause for a bit to allow the system metrics monitoring to exit.
+    time.sleep(1)
+    thread_names = [thread.name for thread in threading.enumerate()]
+    # Check the system metrics monitoring thread has exited.
+    assert "SystemMetricsMonitor" not in thread_names
+
+    # Get the last step.
+    metrics_history = mlflow.tracking.MlflowClient().get_metric_history(
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:110]:
```suggestion
    metrics_history = mlflow.MlflowClient().get_metric_history(
```

File: tests/system_metrics/test_system_metrics_logging.py [hunk 8]
```diff
@@ -91,3 +91,52 @@ def test_automatic_system_metrics_monitor():
...
+    # Check the step is correctly resumed.
+    metrics_history = mlflow.tracking.MlflowClient().get_metric_history(
+        run.info.run_id, "system/cpu_utilization_percentage"
+    )
+    assert metrics_history[-1].step > last_step
+
+    # Unset the environment variables to avoid affecting other test cases.
+    mlflow.disable_system_metrics_logging()
+    mlflow.set_system_metrics_sampling_interval(None)
+    mlflow.set_system_metrics_samples_before_logging(None)
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:142]: you might want to add this as an autouse fixture in case an error occurs before we reach these lines.

File: tests/system_metrics/test_system_metrics_logging.py [hunk 9]
```diff
@@ -91,3 +91,52 @@ def test_automatic_system_metrics_monitor():
...
+        "disk_available_megabytes",
+        "network_receive_megabytes",
+        "network_transmit_megabytes",
+    ]
+    expected_metrics_name = [f"system/{name}" for name in expected_metrics_name]
+    for name in expected_metrics_name:
+        assert name in metrics
+
+    # Check the step is correctly resumed.
+    metrics_history = mlflow.tracking.MlflowClient().get_metric_history(
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:134]:
```suggestion
    metrics_history = mlflow.MlflowClient().get_metric_history(
```

File: mlflow/system_metrics/system_metrics_monitor.py [hunk 10]
```diff
@@ -60,8 +69,26 @@ def __init__(self, run_id, sampling_interval=10, samples_before_logging=1):
...
         self._metrics_prefix = "system/"
+        self._logging_step = self._get_next_logging_step(run_id) if resume_logging else 0
+
+    def _get_next_logging_step(self, run_id):
+        from mlflow.tracking.client import MlflowClient
+
+        client = MlflowClient()
+        try:
+            run = client.get_run(run_id)
+        except Exception:
```
Comment (harupy) [inline mlflow/system_metrics/system_metrics_monitor.py:81]: `Exception` seems too broad. Can we catch `MlflowException`?

File: mlflow/system_metrics/system_metrics_monitor.py [hunk 11]
```diff
@@ -60,8 +69,26 @@ def __init__(self, run_id, sampling_interval=10, samples_before_logging=1):
...
+            return 0
+        system_metric_name = None
+        for metric_name in run.data.metrics.keys():
+            if metric_name.startswith(self._metrics_prefix):
+                system_metric_name = metric_name
+                break
+        if system_metric_name is None:
+            return 0
```
Comment (harupy) [inline mlflow/system_metrics/system_metrics_monitor.py:89]: We can use `next` here:

```python
    system_metric_name = next(
        (m for m in run.data.metrics if m.startswith(self._metrics_prefix)), None
    )
    if system_metric_name is None:
        return 0
```

File: tests/system_metrics/test_system_metrics_logging.py [hunk 12]
```diff
@@ -91,3 +91,52 @@ def test_automatic_system_metrics_monitor():
...
+        "disk_usage_percentage",
+        "disk_usage_megabytes",
+        "disk_available_megabytes",
+        "network_receive_megabytes",
+        "network_transmit_megabytes",
+    ]
+    expected_metrics_name = [f"system/{name}" for name in expected_metrics_name]
+    for name in expected_metrics_name:
+        assert name in metrics
+
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:132]:
```suggestion
    expected_metric_names = [f"system/{name}" for name in expected_metrics_name]
    assert sorted(metrics) == expected_metrics_name

```

File: mlflow/tracking/fluent.py [hunk 13]
```diff
@@ -376,32 +376,32 @@ def start_run(
...
-            try:
-                from mlflow.system_metrics.system_metrics_monitor import SystemMetricsMonitor
+    if log_system_metrics is None:
+        # If `log_system_metrics` is not specified, we will check environment variable.
+        log_system_metrics = MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING.get()
+
+    if log_system_metrics:
+        if importlib.util.find_spec("psutil") is None:
+            raise MlflowException(
+                "Failed to start system metrics monitoring as package `psutil` is not installed."
```
Comment (harupy) [inline mlflow/tracking/fluent.py:387]:
```suggestion
                "Failed to start system metrics monitoring as package `psutil` is not installed ."
```

File: tests/system_metrics/test_system_metrics_logging.py (hunk 12 above)
Comment (chenmoneygithub) [reply tests/system_metrics/test_system_metrics_logging.py:132]: it's not guaranteed to be equivalent. The expected set does not contain GPU metrics, which is logged when GPU is available.

File: tests/system_metrics/test_system_metrics_logging.py [hunk 14]
```diff
@@ -91,3 +91,52 @@ def test_automatic_system_metrics_monitor():
...
+def test_automatic_system_metrics_monitor_resume_existing_run():
+    mlflow.enable_system_metrics_logging()
+    mlflow.set_system_metrics_sampling_interval(0.2)
+    mlflow.set_system_metrics_samples_before_logging(2)
+    with mlflow.start_run() as run:
+        time.sleep(2)
+
+    # Pause for a bit to allow the system metrics monitoring to exit.
+    time.sleep(1)
+    thread_names = [thread.name for thread in threading.enumerate()]
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply tests/system_metrics/test_system_metrics_logging.py:105]: I kinda prefer to align the testing code with users' code, for which they won't explicitly join threads returned by `threading.enumerate`

File: tests/system_metrics/test_system_metrics_logging.py [hunk 15]
```diff
@@ -91,3 +91,52 @@ def test_automatic_system_metrics_monitor():
...
+    ]
+    expected_metrics_name = [f"system/{name}" for name in expected_metrics_name]
+    for name in expected_metrics_name:
+        assert name in metrics
+
+    # Check the step is correctly resumed.
+    metrics_history = mlflow.tracking.MlflowClient().get_metric_history(
+        run.info.run_id, "system/cpu_utilization_percentage"
+    )
+    assert metrics_history[-1].step > last_step
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply tests/system_metrics/test_system_metrics_logging.py:137]: the behavior here is not deterministic because of python thread management.

File: mlflow/system_metrics/system_metrics_monitor.py (hunk 11 above)
Comment (chenmoneygithub) [reply mlflow/system_metrics/system_metrics_monitor.py:89]: I think the original code is easier to read? 

`(m for m in run.data.metrics if m.startswith(self._metrics_prefix)), None` inlines multiple things.

File: mlflow/system_metrics/system_metrics_monitor.py (hunk 10 above)
Comment (chenmoneygithub) [reply mlflow/system_metrics/system_metrics_monitor.py:81]: sg!

File: tests/system_metrics/test_system_metrics_logging.py (hunk 8 above)
Comment (chenmoneygithub) [reply tests/system_metrics/test_system_metrics_logging.py:142]: nice call.

File: tests/system_metrics/test_system_metrics_logging.py (hunk 12 above)
Comment (harupy) [reply tests/system_metrics/test_system_metrics_logging.py:138]: got it

File: mlflow/system_metrics/system_metrics_monitor.py (hunk 11 above)
Comment (harupy) [reply mlflow/system_metrics/system_metrics_monitor.py:90]: agree

File: tests/system_metrics/test_system_metrics_logging.py [hunk 16]
```diff
@@ -1,10 +1,21 @@
...
 import time
 
+import pytest
+
 import mlflow
 from mlflow.system_metrics.system_metrics_monitor import SystemMetricsMonitor
 
 
+@pytest.fixture(scope="module", autouse=True)
+def clean_mlruns_dir():
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:11]:
```suggestion
def clean_mlruns_dir():
```

Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:11]: can we rename this fixture?

Comment (harupy) [reply tests/system_metrics/test_system_metrics_logging.py:11]: should the scope be module? should we run this after each test?

File: tests/system_metrics/test_system_metrics_logging.py (hunk 12 above)
Comment (harupy) [reply tests/system_metrics/test_system_metrics_logging.py:138]:
```
assert metrics.keys() >= set(expected_metrics_name)
```

might work.

File: tests/system_metrics/test_system_metrics_logging.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (harupy) [reply tests/system_metrics/test_system_metrics_logging.py:143]: what's not deterministic? The number of steps?

**Review by harupy — APPROVED** [review]
LGTM

File: tests/system_metrics/test_system_metrics_logging.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply tests/system_metrics/test_system_metrics_logging.py:143]: yes, the number of steps is not guaranteed.

File: tests/system_metrics/test_system_metrics_logging.py (hunk 16 above)
Comment (chenmoneygithub) [reply tests/system_metrics/test_system_metrics_logging.py:11]: sorry my bad.
